frontend/controller/service-worker.js

```diff
@@ -13,23 +15,95 @@ sbp('sbp/selectors/register', {
     // TODO: move ahead with encryption stuff ignoring this service worker stuff for now
     // TODO: improve updating the sw: https://stackoverflow.com/a/49748437
     // NOTE: user should still be able to use app even if all the SW stuff fails
+    if (!('serviceWorker' in navigator)) { return }
+
     try {
-      await navigator.serviceWorker.register('/assets/js/sw-primary.js', {
+      const registration = await navigator.serviceWorker.register('/assets/js/sw-primary.js', {
         scope: '/'
       })
-      navigator.serviceWorker.addEventListener('message', function (event) {
-        console.info('Received message from SW:', event.data)
-      })
-      setTimeout(function () {
-        navigator.serviceWorker.controller.postMessage({
-          type: 'set',
-          key: 'foo',
-          value: 'bar'
+
+      if (Notification.permission !== 'granted') {
+        // To be able to receive a notification, this setting must be turned on.
+        // But user can customise this setting later via 'user-settings' page.
+        await requestNotificationPermission(true)
+      }
+
+      const API_URL = sbp('okTurtles.data/get', 'API_URL')
+      const existingSubscription = await registration.pushManager.getSubscription()
+
+      if (existingSubscription) {
+        // If there is an existing subscription, no need to create a new one.
+        // But make sure server knows the subscription details too.
+        await fetch(`${API_URL}/push/subscribe`, { method: 'POST', body: JSON.stringify(existingSubscription.toJSON()) })
+        return
+      } else {
+        // Create a new push subscription
+
+        // 1. Get the public key from the server
+        const PUBLIC_VAPID_KEY = await fetch(`${API_URL}/push/publickey`)
+          .then(handleFetchResult('text'))
+
+        // 2. Add a new subscription to pushManager using it.
+        const subscription = await registration.pushManager.subscribe({
```

Comment:
ditto

Reply:
`registration` here is a instance created by `navigator.serviceWorker.register()` in above line. So I believe checking `'serviceWorker' in navigator` earlier will suffice.